### DISCUSSION TITLE
Is there a way to constrain `z.object` to values of an interface?

### DISCUSSION THREAD

**Discussion by luke-h1** [colinhacks/zod#2391] [Q&A] [answered]
Hey, I'm looking at moving over to zod (from `yup`) and am facing a problem when trying to constrain the types allowed in my validationSchema zod object. I'm wondering if it's possible to have an interface such as this: 

```typescript 
interface FormValues {
  email: string;
  password: string;
}
```
and then be able to constrain the `validationSchema` I create to just the two fields in `FormValues` i.e:

```typescript
// only email and password are allowed in validationSchema
const validationSchema = z.object<FormValues>({
  email: z.string({
    required_error: 'Enter an email address',
  }).email(),
  password: z.string({
    required_error: 'Enter a password',
  })
})

```
At the moment TypeScript complains about the `FormValues` interface not satisfying the constraint of `ZodRawShape`

**Comment by MikeRippon** [comment, marked as answer]
Zod is designed for the reverse. You're not expected to define the FormValues interface and try to make the zod schema conform to it, instead you can infer the type automatically from the zod schema:

```typescript
type FormValues = z.infer<typeof validationSchema>
```

That bugged me for a while because I was reluctant to couple my system-wide typings to a third-party library, but in many circumstances it's really nice to just declare a bunch of properties once and avoid the duplication/coupling of having to declare all your properties twice (in schema and type).

_Having said that_, if you're planning on using zod for validation at the API level, there's a strong possibility you don't actually want to couple those things. You should consider whether you would benefit from having separate types for your domain model and API request bodies to allow for differences & to provide a flex-point for future changes. For example:

```typescript
type Person = {
  name: string
  dateOfBirth: Date
  age: number
  optInToNewFeature?: boolean // Optional because there is legacy data in the database from before feature existed
}

const CreatePersonBody = z.object({
  name: z.string(),
  dateOfBirth: z.date(),
  optInToNewFeature: z.boolean(), // Not optional for new API requests
})

function handleCreatePersonWebRequest() {
  const body = CreatePersonSchema.parse(requestBody)

  // Structural typing rocks
  storePersonInDb({
    ...body,
    age: //calculate from body.dateOfBirth
  })
}
```

If like me, you're coming from a name-based language such as Java, that's a mind-blowingly small amount of code to decouple your API and domain models.

**Reply by geoyws** [reply]
Unfortunately most of the time in our work we have to use interfaces given to us from libraries and we cannot expect these teams to create their interfaces from Zod. Could we support object shape matching from TS interfaces when declaring Zod Objects perhaps via a generic? (Apologies I'm a first day Zod user and was expecting this to have worked out of the box.)

**Comment by jvnlwn** [comment]
You can use the `z.ZodSchema` generic:

```typescript
interface FormValues {
  email: string
  password: string
}

const validationSchema: z.ZodSchema<FormValues> = z.object({
  email: z
    .string({
      required_error: "Enter an email address"
    })
    .email(),
  password: z.string({
    required_error: "Enter a password"
  })
})
```

**Reply by geoyws** [reply]
But this doesn't allow us access into `validationSchema.shape`. Anyone know any other way to tie an Interface to a ZodObject?

**Reply by MikeRippon** [reply]
@geoyws that's because the type declaration `:z.ZodSchema<FormValues>` effectively erases the inferred type from `z.object`.

If you're using typescript 4.9+ you can use the [satisfies](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-9.html) keyword to force a type check without losing the inferred type.

```typescript
interface FormValues {
    email: string
    password: string
}

export const validationSchema = z.object({
    email: z
        .string({
            required_error: "Enter an email address"
        })
        .email(),
    password: z.string({
        required_error: "Enter a password"
    })
}) satisfies z.ZodSchema<FormValues>

```

**Reply by BrianMaldo** [reply]
This is how you do it! Thanks!
